[AArch64][llvm] Armv9.7-A: Add support for SVE2p3 LUTI6 operations

jthackray opened the pull request:
Add instructions for SVE2p3 LUTI6 operations:
  - LUTI6 (16-bit)
  - LUTI6 (8-bit)
  - LUTI6 (vector, 16-bit)
  - LUTI6 (table, four registers, 8-bit)
  - LUTI6 (table, single, 8-bit)

as documented here:

  * https://developer.arm.com/documentation/ddi0602/2025-09/
  * https://developer.arm.com/documentation/109697/2025_09/2025-Architecture-Extensions
  
Co-authored-by: Virginia Cangelosi <virginia.cangelosi@arm.com>

jthackray committed:
[AArch64][llvm] Armv9.7-A: Add support for SVE2p3 LUTI6 operations

Add instructions for SVE2p3 LUTI6 operations:
  - LUTI6 (16-bit)
  - LUTI6 (8-bit)
  - LUTI6 (vector, 16-bit)
  - LUTI6 (table, four registers, 8-bit)
  - LUTI6 (table, single, 8-bit)

as documented here:

  * https://developer.arm.com/documentation/ddi0602/2025-09/
  * https://developer.arm.com/documentation/109697/2025_09/2025-Architecture-Extensions

CarolineConcatto commented on llvm/lib/Target/AArch64/AArch64SVEInstrInfo.td:
```diff
@@ -4659,8 +4659,17 @@ let Predicates = [HasSVE2p3_or_SME2p3] in {
+// SVE2.3 instructions
+//===----------------------------------------------------------------------===//
+let Predicates = [HasSVE2p3] in {
+  def LUTI6_Z2ZZ : sve2_luti6_vector<ZPR8, ZZ_b, 0b00011, "luti6">;
```
s/LUTI6_Z2ZZ/LUTI6_Z2ZZ_B/g

CarolineConcatto commented on llvm/test/MC/AArch64/SVE2p3/luti6-diagnostics.s:
```diff
@@ -0,0 +1,70 @@
+// --------------------------------------------------------------------------//
+// Negative tests for instructions that are incompatible with movprfx
+
+movprfx z0.h, p0/m, z7.h
```
Can we have something like this instead:
movprfx z0.h, p0/m, z7.h
luti6 z0.b, { z2.b, z3.b }, z4

jthackray replied:
Thanks, done.

CarolineConcatto commented on llvm/test/MC/AArch64/SVE2p3/luti6-diagnostics.s:
```diff
@@ -0,0 +1,70 @@
+// --------------------------------------------------------------------------//
+// Negative tests for instructions that are incompatible with movprfx
+
+movprfx z0.h, p0/m, z7.h
```
Can we have something like this instead:
movprfx z0.h, p0/m, z7.h
luti6 z0.b, { z2.b, z3.b }, z4[0]

jthackray commented on llvm/lib/Target/AArch64/AArch64SVEInstrInfo.td:
```diff
@@ -4659,8 +4659,17 @@ let Predicates = [HasSVE2p3_or_SME2p3] in {
+// SVE2.3 instructions
+//===----------------------------------------------------------------------===//
+let Predicates = [HasSVE2p3] in {
+  def LUTI6_Z2ZZ : sve2_luti6_vector<ZPR8, ZZ_b, 0b00011, "luti6">;
```
Thanks, good spot. Fixed.

jthackray commented on llvm/test/MC/AArch64/SVE2p3/luti6-diagnostics.s:
```diff
@@ -0,0 +1,70 @@
+// CHECK-NEXT: luti6 z10.h, { z0.h, z1.h }, z0[0]
+// CHECK-NOT: [[@LINE-1]]:{{[0-9]+}}:
+
+movprfx z0, z7
```
Done

jthackray commented on llvm/test/MC/AArch64/SVE2p3/luti6-diagnostics.s:
```diff
@@ -0,0 +1,70 @@
+// CHECK-NEXT: luti6 z10.b, { z0.b, z1.b }, z0
+// CHECK-NOT: [[@LINE-1]]:{{[0-9]+}}:
+
+movprfx z0, z7
```
Done

jthackray committed:
fixup! [AArch64][llvm] Armv9.7-A: Add support for SVE2p3 LUTI6 operations

Test movprfx properly and remove parameters from sve2_luti6_vector
since it's only used once.

jthackray committed:
fixup! [AArch64][llvm] Armv9.7-A: Add support for SVE2p3 LUTI6 operations

The code in tryParseVectorList() should only apply to `luti6` instructions

jthackray committed:
fixup! [AArch64][llvm] Armv9.7-A: Add support for SVE2p3 LUTI6 operations

Add extra `luti6` tests that should be rejected

jthackray committed:
fixup! [AArch64][llvm] Armv9.7-A: Add support for SVE2p3 LUTI6 operations

Use base class to share more of the encodings for luti6

CarolineConcatto reviewed (commented):
I am  sorry Jonathan I missed the changes in the parser last time.

jthackray committed:
Revert "fixup! [AArch64][llvm] Armv9.7-A: Add support for SVE2p3 LUTI6 operations"

This reverts commit 40c213805546c17b5cd887062ccc030636736c48.

jthackray commented on llvm/lib/Target/AArch64/AsmParser/AArch64AsmParser.cpp:
```diff
@@ -4882,6 +4882,15 @@ ParseStatus AArch64AsmParser::tryParseVectorList(OperandVector &Operands,
       FirstReg, Count, Stride, NumElements, ElementWidth, VectorKind, S,
       getLoc(), getContext()));
 
+  if (Mnemonic == "luti6") {
```
Ok, I've removed this now.

jthackray committed:
fixup! [AArch64][llvm] Armv9.7-A: Add support for SVE2p3 LUTI6 operations"

Use `is(AsmToken::LBrac)` when parsing indexed luti6 instructions
of the form: `luti6 { z0.h - z3.h }, { z0.h, z1.h }, { z0, z1 }[0]`

CarolineConcatto reviewed (approved):
Thank you Jonathan for addressing the comments.
LGTM!

jthackray commented:
### Merge activity

* **Oct 23, 12:32 PM UTC**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/llvm/llvm-project/163164).
